binder/src/main/java/io/grpc/binder/util/ServiceBinding.java

```diff
@@ -0,0 +1,213 @@
+/*
+ * Copyright 2020 The gRPC Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.grpc.binder.util;
+
+import android.content.ComponentName;
+import android.content.Context;
+import android.content.Intent;
+import android.content.ServiceConnection;
+import android.os.IBinder;
+import androidx.annotation.AnyThread;
+import androidx.annotation.MainThread;
+import com.google.common.annotations.VisibleForTesting;
+import io.grpc.Status;
+import java.util.concurrent.Executor;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+import javax.annotation.Nullable;
+import javax.annotation.concurrent.GuardedBy;
+import javax.annotation.concurrent.ThreadSafe;
+
+/**
+ * Manages an Android binding that's restricted to at most one connection to the remote Service.
+ *
+ * <p>A note on synchronization & locking in this class. Clients of this class are likely to manage
+ * their own internal state via synchronization. In order to avoid deadlocks, we must not hold any
+ * locks while calling observer callbacks.
+ *
+ * <p>For this reason, while internal consistency is handled with synchronization (the state field),
+ * consistency on our observer callbacks is ensured by doing everything on the application's main
+ * thread.
+ */
+@ThreadSafe
+final class ServiceBinding implements Bindable, ServiceConnection {
+
+  private static final Logger logger = Logger.getLogger(ServiceBinding.class.getName());
+
+  // States can only ever transition in one direction.
+  private enum State {
+    NOT_BINDING,
+    BINDING,
+    BOUND,
+    UNBOUND,
+  }
+
+  private final ComponentName targetComponent;
+  private final String bindAction;
+  private final int bindFlags;
+  private final Observer observer;
+  private final Executor mainThreadExecutor;
+
+  @GuardedBy("this")
+  private State state;
+
+  // The following fields are intentionally not guarded, since (aside from the constructor),
+  // they're only used on the main thread. The constructor contains a synchronized block
```

Comment:
We don't want to rely on safe publication?

Reply:
Yeah, in hindsight, that would have been saner. I think it's worth a full cleanup of the threading in a follow up.